packages/cactus-plugin-keychain-google-sm/src/main/typescript/plugin-keychain-google-sm.ts

```diff
@@ -0,0 +1,182 @@
+    const label = this.className;
+    this.log = LoggerProvider.getOrCreate({ level, label });
+
+    this.log.info(`Created ${this.className}. KeychainID=${opts.keychainId}`);
+  }
+
+  async registerWebServices(app: Express): Promise<IWebServiceEndpoint[]> {
+    const webServices = await this.getOrCreateWebServices();
+    await Promise.all(webServices.map((ws) => ws.registerExpress(app)));
+    return webServices;
+  }
+
+  public async getOrCreateWebServices(): Promise<IWebServiceEndpoint[]> {
+    if (Array.isArray(this.endpoints)) {
+      return this.endpoints;
+    }
+    const endpoints: IWebServiceEndpoint[] = [];
+
+    // TODO: Writing the getExpressRequestHandler() method for
+    // GetKeychainEntryEndpointV1 and SetKeychainEntryEndpointV1
+
+    // {
+    //   const ep = new GetKeychainEntryEndpointV1({
+    //     logLevel: this.opts.logLevel,
+    //   });
+    //   ep.registerExpress(expressApp);
+    //   endpoints.push(ep);
+    // }
+    // {
+    //   const ep = new SetKeychainEntryEndpointV1({
+    //     logLevel: this.opts.logLevel,
+    //   });
+    //   ep.registerExpress(expressApp);
+    //   endpoints.push(ep);
+    // }
+    this.endpoints = endpoints;
+
+    return endpoints;
+  }
+
+  public getHttpServer(): Optional<Server | SecureServer> {
+    return Optional.empty();
+  }
+
+  public async shutdown(): Promise<void> {
+    throw new Error("Method not implemented.");
+  }
+
+  public getInstanceId(): string {
+    return this.instanceId;
+  }
+
+  public getKeychainId(): string {
+    return this.opts.keychainId;
+  }
+
+  public getPackageName(): string {
+    return `@hyperledger/cactus-plugin-keychain-vault`;
+  }
+
+  async rotateEncryptionKeys(): Promise<void> {
+    throw new Error("Method not implemented.");
+  }
+
+  public getEncryptionAlgorithm(): string {
+    return null as any;
+  }
+
+  public getGoogleSmClient(): SecretManagerServiceClient {
+    return this.googleSmClient;
+  }
+
+  async get<T>(key: string): Promise<T> {
+    const accessResponse = await this.googleSmClient.getSecret({
+      name: key,
+    });
+    if (accessResponse[0]) {
+      const result = accessResponse[0];
+      return (result as unknown) as T;
+    } else {
+      throw new Error(`${key} secret not found.`);
+    }
+  }
+
+  /**
+   * Detects the presence of a key by trying to read it and then
+   * observing whether an HTTP 404 NOT FOUND error is returned or
+   * not and deciding whether the keychain has the entry ot not
+   * based on this.
+   */
+  async has(key: string): Promise<boolean> {
+    const accessResponse = await this.googleSmClient.accessSecretVersion({
+      name: key,
+    });
+    if (accessResponse) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  async set<T>(key: string, value: T): Promise<void> {
+    const parent = key.split("?")[0];
+    const secretId = key.split("?")[1];
+    const secret = await this.googleSmClient.createSecret({
```

Comment:
@jagpreetsinghsasan Is there some logic here that we need to take care of so that the set method is re-entrant?

Reply:
I added the comment in test case, but forgot to mention here.
Basically the key should consist of parent and secretId seperated by ?
For example, key = "projects/my-project?my-secret"

I have added that now.

Reply:
@jagpreetsinghsasan What I mean: would this code work or would the second invocation crash? (If it crashes it's not re-entrant, if it doesn't crash, then it is)

```typescript
await keychain.set("x?y", "something");
await keychain.set("x?y", "something");
```

Reply:
It works for the second time too, so I believe that this method is re-entrant